### PR TITLE
Inscribe png and text

### PR DESCRIPTION
**Pull request by raphjaph**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -33,7 +34,7 @@ log = "0.4.14"
 mime = "0.3.16"
 mime_guess = "2.0.4"
 ord-bitcoincore-rpc = "0.16.3"
-redb = "0.8.0"
+redb = "0.9.0"
```
Comment (casey) [inline Cargo.toml:37]: We aren't using tuple types yet, so we can go back to redb 0.8.0.

File: src/index.rs [hunk 2]
```diff
@@ -339,7 +339,7 @@ impl Index {
         .begin_read()?
         .open_table(ORDINAL_TO_INSCRIPTION)?
         .get(&ordinal.n())?
-        .map(|inscription| Inscription(inscription.to_owned())),
```
Comment (casey) [inline src/index.rs:342]: Add context and return the error. Something like "Failed to deserialize inscription".

File: src/index/updater.rs [hunk 3]
```diff
@@ -291,7 +291,12 @@ impl Updater {
     outputs_traversed: &mut u64,
   ) -> Result {
     if let Some((ordinal, inscription)) = Inscription::from_transaction(tx, input_ordinal_ranges) {
-      ordinal_to_inscription.insert(&ordinal.n(), &inscription.0)?;
+      // let json = match inscription {
+      // Inscription::Text(content) => serde_json::json!({ "media_type": "text/plain;charset=utf-8", "content": content,}),
+      // Inscription::Png(content) => serde_json::json!({ "media_type": "image/png", "content": content,}),
+      // };
```
Comment (casey) [inline src/index/updater.rs:297]:
```suggestion
```

File: src/index/updater.rs [hunk 4]
```diff
@@ -291,7 +291,12 @@ impl Updater {
...
+      let json = serde_json::to_string(&inscription).unwrap();
```
Comment (casey) [inline src/index/updater.rs:298]: I think this unwrap is fine. We should use an `expect` though. "Inscription serialization should always succeed"

**Review by casey — CHANGES_REQUESTED** [review]
brain not work

File: src/inscription.rs [hunk 5]
```diff
@@ -99,19 +102,41 @@ impl<'a> InscriptionParser<'a> {
...
-      let content = if let Instruction::PushBytes(bytes) = content {
+      let media_type = if let Instruction::PushBytes(bytes) = media_type {
         str::from_utf8(bytes).map_err(InscriptionError::Utf8Decode)?
       } else {
         return Err(InscriptionError::InvalidInscription);
       };
 
+      let inscription = match media_type {
+        "text/plain;charset=utf-8" => {
+          let content = if let Instruction::PushBytes(bytes) = self.advance()? {
```
Comment (casey) [inline src/inscription.rs:115]: Move advance outside of the match, and convert to `Vec<u8>`. (There is a Vec<u8> -> String conversion you can use if it's text.

File: src/inscription.rs [hunk 6]
```diff
@@ -99,19 +102,41 @@ impl<'a> InscriptionParser<'a> {
 
   fn parse_inscription(&mut self) -> Result<Option<Inscription>> {
     if self.advance()? == Instruction::Op(opcodes::all::OP_IF) {
-      let content = self.advance()?;
+      let media_type = self.advance()?;
 
-      let content = if let Instruction::PushBytes(bytes) = content {
+      let media_type = if let Instruction::PushBytes(bytes) = media_type {
```
Comment (casey) [inline src/inscription.rs:107]:
```suggestion
      let media_type = if let Instruction::PushBytes(bytes) = self.advance()? {
```

File: src/inscription.rs [hunk 7]
```diff
@@ -167,13 +192,14 @@ mod tests {
...
       .push_opcode(opcodes::all::OP_IF)
+      .push_slice("text/plain;charset=utf-8".as_bytes())
       .push_slice("ord".as_bytes())
       .push_opcode(opcodes::all::OP_ENDIF)
       .into_script();
 
     assert_eq!(
       InscriptionParser::parse(&Witness::from_vec(vec![script.into_bytes(), vec![]])),
-      Ok(Inscription("ord".into()))
+      Ok(Inscription::Text("ord".to_string()))
```
Comment (casey) [inline src/inscription.rs:202]:
```suggestion
      Ok(Inscription::Text("ord".into()))
```

File: src/inscription.rs [hunk 8]
```diff
@@ -182,14 +208,15 @@ mod tests {
...
+      .push_slice("text/plain;charset=utf-8".as_bytes())
       .push_slice("ord".as_bytes())
       .push_opcode(opcodes::all::OP_ENDIF)
       .push_opcode(opcodes::all::OP_CHECKSIG)
       .into_script();
 
     assert_eq!(
       InscriptionParser::parse(&Witness::from_vec(vec![script.into_bytes(), vec![]])),
-      Ok(Inscription("ord".into()))
+      Ok(Inscription::Text("ord".to_string()))
```
Comment (casey) [inline src/inscription.rs:219]:
```suggestion
      Ok(Inscription::Text("ord".into()))
```

File: src/inscription.rs [hunk 9]
```diff
@@ -313,7 +345,7 @@ mod tests {
 
     assert_eq!(
       Inscription::from_transaction(&tx, &ranges),
-      Some((Ordinal(1), Inscription("ord".into())))
+      Some((Ordinal(1), Inscription::Text("ord".into())))
     );
   }
```
Comment (casey) [inline src/inscription.rs:351]: Add a test that creates an `Inscription::Png`

File: src/subcommand/server/templates/ordinal.rs [hunk 10]
```diff
@@ -85,7 +85,7 @@ mod tests {
       OrdinalHtml {
         ordinal: Ordinal(0),
         blocktime: Blocktime::Confirmed(0),
-        inscription: Some(Inscription("HELLOWORLD".to_string())),
+        inscription: Some(Inscription::Text("HELLOWORLD".to_string())),
```
Comment (casey) [inline src/subcommand/server/templates/ordinal.rs:88]:
```suggestion
        inscription: Some(Inscription::Text("HELLOWORLD".into())),
```

File: src/subcommand/server/templates/ordinal.rs [hunk 11]
```diff
@@ -117,7 +117,7 @@ mod tests {
       OrdinalHtml {
         ordinal: Ordinal(0),
         blocktime: Blocktime::Confirmed(0),
-        inscription: Some(Inscription(
+        inscription: Some(Inscription::Text(
           "<script>alert('HELLOWORLD');</script>".to_string()
```
Comment (casey) [inline src/subcommand/server/templates/ordinal.rs:121]:
```suggestion
          "<script>alert('HELLOWORLD');</script>".into()
```

File: src/subcommand/wallet/inscribe.rs [hunk 12]
```diff
@@ -10,18 +10,33 @@ use {
...
 
 #[derive(Debug, Parser)]
 pub(crate) struct Inscribe {
+  #[clap(long, help = "Inscribe on <ORDINAL>")]
   ordinal: Ordinal,
-  content: String,
+  #[clap(long, help = "Inscribe contents of this <FILE>")]
+  content: PathBuf,
+  #[clap(long, help = "Set the media type of the contents")]
+  media_type: Mime,
```
Comment (casey) [inline src/subcommand/wallet/inscribe.rs:23]:
```suggestion
  #[clap(long, help = "Inscribe content of <FILE>")]
  file: PathBuf,
```

File: src/subcommand/wallet/inscribe.rs [hunk 13]
```diff
@@ -59,6 +75,7 @@ impl Inscribe {
 
   fn create_inscription_transactions(
     ordinal: Ordinal,
+    media_type: Mime,
     content: &[u8],
```
Comment (casey) [inline src/subcommand/wallet/inscribe.rs:79]:
```suggestion
  inscription: Inscription,
```

File: src/inscription.rs [hunk 14]
```diff
@@ -11,8 +11,11 @@ use {
...
 
-#[derive(Debug, PartialEq)]
-pub(crate) struct Inscription(pub(crate) String);
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub(crate) enum Inscription {
+  Text(String),
+  Png(Vec<u8>),
+}
 
 impl Inscription {
```
Comment (casey) [inline src/inscription.rs:20]:
```suggestion
impl Inscription {
  fn media_type(&self) -> &str {
  }
```

File: templates/ordinal.html [hunk 15]
```diff
@@ -11,9 +11,15 @@ <h1>Ordinal {{ self.ordinal.n() }}</h1>
   <dt>offset</dt><dd>{{ self.ordinal.third() }}</dd>
   <dt>rarity</dt><dd><span class={{self.ordinal.rarity()}}>{{ self.ordinal.rarity() }}</span></dd>
   <dt>time</dt><dd>{{ self.blocktime }}</dd>
-%% if let Some(Inscription(inscription)) = &self.inscription {
-  <dt>inscription</dt><dd>{{inscription}}</dd>
-%% }
+%% match &self.inscription {
```
Comment (casey) [inline templates/ordinal.html:14]:
```suggestion
%% if let Some(inscription) = &self.inscription {
%% match inscription {
```

File: templates/ordinal.html [hunk 16]
```diff
@@ -11,9 +11,15 @@ <h1>Ordinal {{ self.ordinal.n() }}</h1>
...
+%%  Some(Inscription::Text(content)) => {
+  <dt>inscription</dt><dd>{{ content }}</dd>
+%% },
+%%  Some(Inscription::Png(content)) => {
+  <dt>inscription</dt><dd><img src="data:image/png;base64,{{ base64::encode(&content) }}"></dd>
```
Comment (casey) [inline templates/ordinal.html:19]: Confirm visually that this works.

File: tests/wallet.rs [hunk 17]
```diff
@@ -185,9 +186,48 @@ fn inscribe_forbidden_on_mainnet() {
...
+    "wallet inscribe --ordinal 5000000000 --content hello.txt --media-type text/plain;charset=utf-8",
+  )
+  .rpc_server(&rpc_server)
+  .expected_exit_code(1)
+  .expected_stderr("error: `ord wallet inscribe` is unstable and not yet supported on mainnet.\n")
+  .run();
+}
+
+#[test]
+fn inscribe_invalid_mime_type() {
```
Comment (casey) [inline tests/wallet.rs:199]: Inscribe unknown file extension

File: tests/wallet.rs [hunk 18]
```diff
@@ -185,9 +186,48 @@ fn inscribe_forbidden_on_mainnet() {
...
+  .expected_stderr("error: inscribe only accepts text/plain;charset=utf-8 and image/png\n")
+  .run();
+}
+
+#[test]
+fn inscribe_png() {
+  let rpc_server = test_bitcoincore_rpc::spawn_with(Network::Regtest, "ord");
+  rpc_server.mine_blocks(1);
+
+  let png = base64::decode("iVBORw0KGgoAAAANSUhEUgAAAEEAAABBCAYAAACO98lFAAACbElEQVR4nM2bW3LCQAwEY+5/5wSRWsos+5A0LUNXJQR7Lc108cNHjt87PwGO47j//m6ClX5u958Q0QVXk8kXlmBkFl1BNldKgpFdWIWSJy3BUBaTqDkkCYYaQIXYL0swiCAZqL2IBIMK5IXch0kwyGAr6D2oBIMO2FMxH5dgVAQ1quaWSDDowPS8M7fK7wJUcGrOCOv/+CTYH1WoBdTnV7TeDwlGu1BBtkj2OQ/nvk8JxvkGTbRQ9HyEvueLBKM/QOIt5j2XYdTvTYIxOkixK7i7rzDrNZRgzB4gmBWdXSdY9ZlKMFYPqvSF+/ckux5LCcZugEIr3l4r8OS3E64ElUGr8Agwtp+EhnfgtxDJ65ZgRAZ/kmjOkAQjuuBqMvnCEozMoivI5kpJMLILq1DypCUYymISNYckwVADqBD7ZQkGESQDtReRYFCBvJD7MAkGGWwFvQeVYNABeyrm4xKqv2NUzEclVAQcQe/BJNDBdpD7EAlkoAjUXlkCFSQLsV+SQAQgUHOkJaiLaZQ8KQnKwkqyucISsouuIpMvJCGz4BNEc7olRAd/mkhel4TIwCgV3wUa3txbCd5BGZqA9lqBJ/9SgmdAlr54/55k12MqYfegwqzw7DrBqs9QwuoBlV3R3X2FWa83CbODBN6C3nMZRv1eJIwOUESLRc9H6Hs+JfQ3SLKFss95OPd9SDhfoFGLqM+vaL3D/wgWgSxQGPP/k1ABKcCg550pkVAVuGouLqEqaKNiPiqhIuAIeg8mgQ62g9yHSCADRaD2yhKoIFmI/ZIEIgCBmiMtQV1Mo+RJSVAWVpLNFZaQXXQVmXx/8lkCgwQhAzMAAAAASUVORK5CYII=").unwrap();
```
Comment (casey) [inline tests/wallet.rs:217]: This can just be a `Vec<u8>`

**Review by casey — CHANGES_REQUESTED** [review]

File: src/inscription.rs [hunk 19]
```diff
@@ -25,6 +28,20 @@ impl Inscription {
 
     Some((Ordinal(*start), inscription))
   }
+
+  pub(crate) fn media_type(&self) -> &[u8] {
```
Comment (casey) [inline src/inscription.rs:32]:
```suggestion
  pub(crate) fn media_type(&self) -> &str {
```

File: src/inscription.rs [hunk 20]
```diff
@@ -25,6 +28,20 @@ impl Inscription {
...
+    match self {
+      Inscription::Text(_) => "text/plain;charset=utf-8".as_bytes(),
+      Inscription::Png(_) => "image/png".as_bytes(),
```
Comment (casey) [inline src/inscription.rs:35]:
```suggestion
      Inscription::Text(_) => "text/plain;charset=utf-8",
      Inscription::Png(_) => "image/png",
```

File: src/inscription.rs [hunk 21]
```diff
@@ -99,19 +116,35 @@ impl<'a> InscriptionParser<'a> {
...
+      let inscription = match media_type {
+        "text/plain;charset=utf-8" => Some(Inscription::Text(
+          str::from_utf8(content)
+            .map_err(InscriptionError::Utf8Decode)?
+            .into(),
+        )),
+        "image/png" => Some(Inscription::Png(content.to_vec())),
+        _ => {
+          return Err(InscriptionError::InvalidInscription);
+        }
```
Comment (casey) [inline src/inscription.rs:140]: I think we should have an unknown inscription type:

```suggestion
        _ => Inscription::Unknown { media_type, content },
```

Comment (casey) [reply src/inscription.rs:140]: This is so, if we later need to have a canonical view of all inscriptions, which inscriptions exist doesn't depend on what mime type they have.

Comment (casey) [reply src/inscription.rs:140]: Also add a test that unknown inscriptions are parsed from scripts correctly.

File: src/subcommand/wallet/inscribe.rs [hunk 22]
```diff
@@ -14,14 +14,33 @@ use {
 
 #[derive(Debug, Parser)]
 pub(crate) struct Inscribe {
+  #[clap(long, help = "Inscribe on <ORDINAL>")]
   ordinal: Ordinal,
-  content: String,
+  #[clap(long, help = "Inscribe contents of <FILE>")]
```
Comment (casey) [inline src/subcommand/wallet/inscribe.rs:19]:
```suggestion
  #[clap(long, help = "Inscribe ordinal with contents of <FILE>")]
```

File: src/subcommand/wallet/inscribe.rs [hunk 23]
```diff
@@ -14,14 +14,33 @@ use {
 
 #[derive(Debug, Parser)]
 pub(crate) struct Inscribe {
+  #[clap(long, help = "Inscribe on <ORDINAL>")]
```
Comment (casey) [inline src/subcommand/wallet/inscribe.rs:17]:
```suggestion
  #[clap(long, help = "Inscribe <ORDINAL>")]
```

File: src/subcommand/wallet/inscribe.rs [hunk 24]
```diff
@@ -14,14 +14,33 @@ use {
...
 }
 
 impl Inscribe {
   pub(crate) fn run(self, options: Options) -> Result {
     let client = options.bitcoin_rpc_client_mainnet_forbidden("ord wallet inscribe")?;
 
+    let inscription = match &self
+      .file
+      .extension()
+      .expect("file should always have an extension")
```
Comment (casey) [inline src/subcommand/wallet/inscribe.rs:30]: Files don't always have extensions, so we should handle this error case. Something like:

```suggestion
      .ok_or_else(|| anyhow!("file must have extension"))?
      .to_str()
      .ok_or_else(|| anyhow!("unrecognized extension"))?
```

File: src/subcommand/wallet/inscribe.rs [hunk 25]
```diff
@@ -14,14 +14,33 @@ use {
...
+      .to_str()
+    {
+      Some("txt") => Inscription::Text(String::from_utf8_lossy(&fs::read(&self.file).unwrap()).into()),
+      Some("png") => Inscription::Png(fs::read(&self.file).unwrap()),
+      _ => {
+        return Err(anyhow!("inscribe only accepts .txt and .png"));
+      }
```
Comment (casey) [inline src/subcommand/wallet/inscribe.rs:37]: `from_utf8_lossy` replaces all invalid UTF-8 with replacement characters. Publishing that is probably not what users want.

```suggestion
      "txt" => Inscription::Text(fs::read_to_string(&self.file)?),
      "png" => Inscription::Png(fs::read(&self.file)?),
      other => bail!("Unrecognized file extension: `.{other}`"),
```

File: src/subcommand/wallet/inscribe.rs [hunk 26]
```diff
@@ -14,14 +14,33 @@ use {
...
+    };
+
+    if fs::read(&self.file).unwrap().len() > 520 {
```
Comment (casey) [inline src/subcommand/wallet/inscribe.rs:40]: This reads the file twice. It's probably better to do checks like this in an `Inscription` constructor. How about `Inscription::from_file` that takes a path, and checks the extension, reads the file, and checks the length there?

**Review by casey — CHANGES_REQUESTED** [review]

File: src/subcommand/wallet/inscribe.rs (hunk 25 above)
Comment (raphjaph) [reply src/subcommand/wallet/inscribe.rs:37]: add what extension accepted
`unrecognized file extension  `.{other}`, only .txt and .png are accepted`

**Review by casey — APPROVED** [review]

**Pull request merged**